### DISCUSSION TITLE
Router V2

### DISCUSSION THREAD

**Discussion by tannerlinsley** [TanStack/router#975] [Ideas]
### Moving away from classes

Classes are great for defining types and functionality in one go, but they're not very friendly to destructuring and compositional/functional APIs. The following changes would unify the router API to be less class-y and more function-y.

- [x] Deprecate `new Router` in favor of `createRouter`
- [x] Deprecate `new RootRoute` in favor of `createRootRoute`
- [x] Deprecate `new Route` in favor of `createRoute`
- [x] Deprecate `new FileRoute(path).createRoute(options)` in favor of `createFileRoute(path)(options)`
- [x] Deprecate `new RouteApi` in favor of `getRouteAPI`
- [x] Deprecate `FileRouteLoader` in favor of the new `createLazyFileRoute` and `Route.lazy` functions

### JSDoc Annotations

- [ ] Add inline JSDoc annotations to most of the API to help with IDE documentation
- [ ] Potentially auto-generate API reference documentation

### Code Splitting 

- [x] Deprecate individual code-split properties `component`, `errorComponent` and `pendingComponent` in favor of `createLazyFileRoute` and `Route.lazy()`
- [ ] Turn on `autoCodeSplitting` by default.

### Code-Generation

- [x] Refactor generated code to be smaller and easier to follow.
- [ ] Move all generated code out of the `src` and `app` directories. Preferably a `.tanstack` directory.

### Server

- [x] Deprecate `escapeJSON` in favor `jsec`.
- [x] Deprecate `dehydrateData` in favor of `injectHtml`.
- [x] Deprecate `hydrateData` in favor of `injectHtml`.

### General

- [x] Deprecate 'Not-Found Routes' in favor of route-specific `notFoundComponent`. See migration guide [here](https://tanstack.com/router/v1/docs/guide/not-found-errors#migrating-from-notfoundroute).
- [x] Deprecate `global` on `NotFoundError` in favor of `routeId`.
- [x] Deprecate `href` in `redirect()` in favor of `statusCode`.
- [x] Deprecate `ErrorRouteProps` in favor of `ErrorComponentProps`.
- [x] Deprecate the `navigate` function as an argument in the beforeLoad and loader callbacks, in favor of using `throw redirect(...)` and `throw notFound(...)`.
- [ ] Remove `errorInfo` from `ErrorComponentProps` ([ref](https://github.com/TanStack/router/issues/2512)).
- [ ] Rename the `key` in `location.state` to `__TSR_key` so that it matches the prefixed naming like `__TSR_index` ([ref](https://github.com/TanStack/router/pull/4356)).

**Comment by SeanCassiere** [comment]
In addition to the migration from classes to the more compositional/function approach, could we ensure consistent naming of all the **user-facing** APIs?
Off the top of my head, the `FileRouteLoader` *function* is one which could be better named as `createFileRouteLoader`.

When a timeline for v2 is more clear we could take stock and compile a list of **user-facing APIs** for renaming.

**Reply by tannerlinsley** [reply]
Added to the list ✅

**Comment by SeanCassiere** [comment]
It's something I saw on X/Twitter a few months back and a variation of which I saw [today](https://twitter.com/antfu7/status/1744696605456962035). It's about the idea of utilizing JSDocs for giving the user documentation right in their code editor. 

## Approach 1
For an easy option, we could do what Tanstack Table did, and add links to the JSDoc comments on each property which would direct the user to the individual headings on the Documentation site.
<details><summary>Screenshot</summary>
<p>
<img width="656" alt="image" src="https://github.com/TanStack/router/assets/33615041/a023c915-be10-41c0-aa4c-00f1914e221c">
</p>
</details> 

 This would be rather a quick task and has a significantly reduced buy-in compared to approach 2.

## Approach 2
We could ship the entire *Reference* docs to the user through JSDocs by setting them on the properties.

It gives the user the benefit of not having to even visit the reference site, whilst retaining the exact version of documentation needed for the package version they are on.

However, this approach would require investigation into how we could extract the comments from JSDocs into usable static documentation, as well as the CI steps that would come with it. Also, on how this would affect bundle sizes.

**Reply by tannerlinsley** [reply]
I've updated the items to include these things :)

**Reply by SeanCassiere** [reply]
Saw this from Ryan today: https://twitter.com/ryanflorence/status/1785440251508572608

**Comment by SeanCassiere** [comment]
Adding a helper function similar to `throw redirect()` to trigger the closest **notFoundRoute** from the `beforeLoad` and `loader` functions.

Use case I'm looking at:
```ts
// /postId
const Route = new Route({
  ...,
  loader: async ({ params: { postId } }) => {
    const post = await getPost(postId);

    if (!post) throw notFound();

    return postl
  }
})
```

The behaviour of this helper during the `preload` state is something to be considered.

**Reply by SeanCassiere** [reply]
Personally, I don't believe this `throw` should do anything during a *preload* state, and should be suppressed (do not trigger the NotFoundRoute).
However, a valid argument could be made that since both the `beforeLoad` and `loader` functions are given the `preload` state as an argument, that it could be left up to the user to decide how they want it to behave.

**Reply by tannerlinsley** [reply]
We have special logic internally to ignore redirects during preloads. It would be very easy to do the same for notFound

**Comment by tannerlinsley** [comment]
I updated this with the work I just did in v1. All of it should be non-breaking.

**Comment by sohamnandi77** [comment]


**Reply by tannerlinsley** [reply]
https://tanstack.com/router/v1/docs/framework/react/guide/code-splitting#using-the-lazytsx-suffix

**Comment by auvansang** [comment]
I'm thinking about decorate LOL, like .NET
@route("/xxxx", ....other args)
export function Component(props) {}

**Reply by SeanCassiere** [reply]
A few things on this:

Since the library is not "Typescript-only", this wouldn't be possible the Javascript decorators proposal is still in Stage 3 of the RFC process and right now has [0% browser availability](https://caniuse.com/decorators) it wouldn't be possible to implement without requiring devs to also add in a polyfill.

...and even when it does come in and if it receives good browser support, it'd then be a question of looking at the state of the JS ecosystem and the user perception of dealing with a new interface like what decorators bring to the table. Since it'll essentially be brand new to the eco-system as a whole (outside of the Nest and TypeORM peeps) people may find it too confusing.

Lastly, and most importantly, since the Typescript part of this Router is all about runtime type-checking and dev type-inference which is a big seller for the library, it remains to be seen if the current level of DX can be implemented using the future official Decorators implementation. Since any regression is the type-inference from the current DX will be a big "no-no".

**TLDR;** I wouldn't hold out too much hope on it *at least* until it has browser implementation. And even when it does it'll be based on whether Tanner sees any *tangible* benefits from adopting it into the codebase.

**Reply by tannerlinsley** [reply]
I can tell you right now that I don’t plan on adopting or supporting decorators unless they somehow become a very very widely used pattern across multiple frameworks.

**Comment by auvansang** [comment]
One more things, I'm adapting to TanStack Router (previously .NextJs), But I still no see a document about routing with Markdown file

**Reply by SeanCassiere** [reply]
Sorry, didn't see this till today... Better late than never I guess.

When it comes to markdown, you'd need a way to fetch and render the markdown in React. Using server rendering, you could fetch the markdown (from a file or a CMS) and pass it to a renderer like react-markdown (or something that does the job of taking markdown and turning it into HTML).

In the aspect of the Router, this is all possible (with server rendering), but you'd need to wire it up to meet your exact use case.

We probably won't go down the route of markdown processing with Router, since that is an implementation detail that can vary quite a bit.

That being said if you are interested in doing so, the [TanStack/tanstack.com](https://github.com/TanStack/tanstack.com) repo is a pretty excellent example of fetching markdown content over the wire and rendering it on the client. You'd probably want to adjust it to read from the file-system instead.

**Comment by mustakimkr** [comment]
Hey @tannerlinsley,

Excited about the upcoming "Render" and "Viewport Visibility" preloading strategies! Considering their importance, I'd like to request early implementation if possible. While I understand there might not be many requests for additional strategies, having more options would offer developers greater flexibility. Could we discuss the possibility of including them in the current version of Router or perhaps the next release?

Thanks for your attention to this!

**Reply by mustakimkr** [reply]
@tannerlinsley seeking your attention!

**Comment by SeanCassiere** [comment]
Exploring the possibility of making the router-plugin compatible with more than just *vite*.

Following up on this [comment](https://github.com/TanStack/router/discussions/1141#discussioncomment-8983101) in a discussion thread, it seems like could be feasible to open up the router-plugin to other build tools.

[unplugin.unjs.io](https://unplugin.unjs.io/guide/) would be the dependency for the compatibility layer, letting us write the plugin once, and get the adapters for vite, rollup, webpack, etc.

The unplugin team, also have an [example plugin](https://github.com/unplugin/unplugin-ast/) using Babel for AST traversal, which would let us bring the automatic code-splitting feature to the other build tools as well.

**Reply by lachieh** [reply]
I've actually got a working example of an unplugin version working here: https://github.com/lachieh/unplugin-tanstack-router

I haven't tested with anything other than vite and rspack, but it _should_ work for all the rest. I just took the same code from the vite plugin and tweaked it in a few places.

**Reply by SeanCassiere** [reply]
@lachieh very cool!

I'll drop this in the maintainers' channel and get the others' thoughts on it.

**Reply by SeanCassiere** [reply]
I'll be tackling this during this week. The current plan is to make the migration and support webpack.

Tracking the migration here: https://github.com/TanStack/router/issues/1756

**Reply by SeanCassiere** [reply]
The TanStack Router plugin has officially been moved over to unplugin.

**Reply by lachieh** [reply]
Nice! Working great. I've archived/deprecated mine and pointed it to the official version. Thanks for the quick turnaround!
https://www.npmjs.com/package/unplugin-tanstack-router

**Reply by SeanCassiere** [reply]
@lachieh thanks for the push! Currently, we've got partial rspack support going! Hopefully, will add more as I get more time 😅.

**Comment by masoodusmani** [comment]
Would love to see parallel routes/named outlets implemented! https://github.com/TanStack/router/discussions/605

**Reply by tannerlinsley** [reply]
Planning it!

**Comment by asnaeb** [comment]
`navigate` function in beforeLoad and loader is actually useful. I use it to redirect on unauthorized requests
```typescript
beforeLoad({navigate}) {
  const client = createClient({
    onRequest() {
      if (auth is expired) {
        navigate({to: "login page"})
      }
    }
  });
  return {client};
}

// Later in some child component

const {client} = Route.useRouteContext();
client.performRequest() // <- Unauthorized request will redirect correctly
```

`throw redirect()` doesn't work in such cases as it isn't always thrown inside a loader. Is there any other way to pass the `navigate` function around that I am not aware of? If not, can we keep it please?

**Reply by SeanCassiere** [reply]
The problem with the `navigate` argument passed into `beforeLoad` and `loader` is that it can cause the app navigations to lag behind. TSR is architected around catching throws of `redirect` and `notFound` and handling those calls accordingly. This especially becomes a problem when people use the `navigate` function in-place of actually using a `redirect` and then complain that it isn't working as expected.

---
For your use-case, could you not create the `client` outside of React (like you are doing in your example) and pass that into the Router's context using the `navigate` method on the `router` instance?

```tsx
const router = createRouter({
  routeTree,
  context: {
    client: undefined! // we'll set this later
  }
})

const client = createClient({
  onRequest() {
    if (auth is expired) {
      router.navigate({to: "login page"})
    }
  }
})

export default function App() {
  return <RouterProvider router={router} context={{ client }} />
}
```

**Reply by asnaeb** [reply]
I've done it this way initially, it gives me errors when reloading nested routes:
`Cannot destructure property '...' of 'Route.useRouteContext(...)' as it is undefined.`;

**Reply by SeanCassiere** [reply]
That shouldn't be happening. Would have to investigated.

Edit: Actually now that I think about it, you don't even need to inject the `client` into the _Router's context_, since you are just using the `client` in the component. You could just directly import it.

**Reply by asnaeb** [reply]
Well, you are actually right. The reason I wanted to inject it is that vite's fast-refresh doesn't like non-component exports. I guess I will live with the export for now. The error remains tho, should I open an issue with a repro?

**Reply by SeanCassiere** [reply]
Provided that you are on the latest version, this may be similar to what's described here: https://github.com/TanStack/router/issues/2010

Also, we've got a new `routeContext` option coming to the route configuration for better stabilization of context and to make it more usable with SSR. Hopefully should ship within the week.
https://github.com/TanStack/router/pull/2104

**Reply by asnaeb** [reply]
Yes, I'm on the latest version and that is the exact issue.

**Comment by BrendanC23** [comment]
It would be helpful to add a callback to `navigate` that would only trigger if the navigation was successful (not blocked by `useBlocker`). See discussion https://github.com/TanStack/router/discussions/271.

**Comment by naoya7076** [comment]
I’m hoping that typed state on a per-route basis will be introduced in v2.
https://github.com/TanStack/router/discussions/284#discussioncomment-9344482

**Comment by tannerlinsley** [comment]
We could totally do that.

Tanner Linsley
TanStack LLC

**Reply by naoya7076** [reply]
I've created a Pull Request to support typed state on a per-route.
https://github.com/TanStack/router/pull/3967

**Comment by jadedevin13** [comment]
I'm hoping that production and dev builds are as much as possible behave the same so that no surprises will happen after prod build.

Listing here the relevant issues:

1. https://github.com/TanStack/router/issues/2238
2. https://github.com/TanStack/router/issues/2241
3. https://github.com/TanStack/router/issues/1967
4. https://github.com/TanStack/router/issues/1866

**Comment by mynk-tmr** [comment]
Can you implement the Form component from React router? I think it saves a lot of imperative coding and the pain of forward refs and typing them in TS ?

**Reply by marcomuser** [reply]
This will be part of react itself with react 19. No need to reimplement this in the router.

**Comment by fkworld** [comment]
Can tsr.config.json options support `routeFileGlob` and `routeFileIgnoreGlob`？

Like [fs.glob](https://nodejs.org/docs/latest/api/fs.html#fspromisesglobpattern-options) or [fast-glob](https://github.com/mrmlnc/fast-glob)

Regex is very difficult to write!

**Reply by alexgusman594-stack** [reply]
fast-glob

**Comment by pavelbelyaev98** [comment]
Would be nice go have an easy way to setup CSP with nounces like nextjs

**Reply by helpisdev** [reply]
Have you found any workaround to this? I tried many things to try and add hashes or nonces to the inline styles generated by react to avoid having to add to the CSP `style-src: 'unsafe-inline'`, but couldn't figure out a way to do it.

**Reply by pavelbelyaev98** [reply]
You dont have to do this. You can safely set "unsafe-inline".

https://scotthelme.co.uk/can-you-get-pwned-with-css/

This is the recommended solution by libraries like headlessUI.
https://github.com/tailwindlabs/headlessui/issues/2615

**Reply by helpisdev** [reply]
Aha, very interesting read. Thanks.

**Comment by mrlubos** [comment]
Hey @tannerlinsley do you plan to eventually make this available for SolidJS?

**Comment by tannerlinsley** [comment]
Whatch Ryan Carniato’s stream tomorrow

**Reply by nivethan-dev** [reply]
you guys are not humans, its a 5 hours stream lol!

**Comment by yslpn** [comment]
Consider adding optional parameters like in react router. This is necessary for painless migration.

https://reactrouter.com/start/declarative/routing#optional-segments

More details is discussion https://github.com/TanStack/router/discussions/146

**Reply by renchap** [reply]
Agreed with this, there is a common pattern to have your main language pages on for example `/about`, and localised version as `/$lang/about`. Optional parameters are probably the best way to do this (alternative would be an optional prefix, if that makes things easier?)

**Reply by Christopher96u** [reply]
This is already merged few days ago :)

**Comment by rafaelcardeallodi** [comment]
Would love to see lazy route tree registration considered for V2 — the ability to dynamically register child routes at navigation time rather than requiring the full route tree before createRouter().
This is a real pain point for Module Federation micro frontend setups, where the host doesn't know remotes' route structures ahead of time and currently has to fetch/register every remote upfront at startup. A beforeLoad-based API to register children on first visit would enable true per-route lazy loading of MFE bundles, cut initial load time, and isolate remote failures to a single route.
I wrote up the use case and a proposed API here: [#7018](https://github.com/TanStack/router/discussions/7018)
